### DISCUSSION TITLE
ip-restriction more specific allow not working

### DISCUSSION THREAD

**Discussion by fnzv** [Kong/kong#7335] [Ideas and feature requests]
Hello there!

### Summary
Recently i was messing around the ip-restrinction plugin and i saw that config.allow & config.deny can work together but only if config.deny is a more specific subnet than config.allow

### Steps To Reproduce

1. Given an example service pointing to localhost (just for the sake of the test, i tried also with some test urls and mockbin.org)
    Configure the rate limit such as:
```
curl -X POST http://127.0.0.1:8001/services/service_example/plugins \
    --data "name=ip-restriction" \
    --data "config.allow=127.0.0.0/24" \
    --data "config.deny=127.0.0.1"
```
2.   Now i try to validate the rule... and is all good because works as expected
 ```
# curl example.com (pointing to localhost)
{
  "message":"Your IP address is not allowed"
}
```
	
3. (After cleaning the old configuration - double checked on konga) I try to invert and make config.deny a more wider subnet than config.allow
```
curl -X POST http://127.0.0.1:8001/services/service_example/plugins \
    --data "name=ip-restriction" \
    --data "config.allow=127.0.0.1" \
    --data "config.deny=0.0.0.0/0"
```

4. Trying to validate the test:
 ```
# curl example.com (pointing to localhost)
{
  "message":"Your IP address is not allowed"
}
```

So it seems to be that the allow rule is being ignored and config.deny has more priority above it

There is an example on the documentation here https://docs.konghq.com/hub/kong-inc/ip-restriction/ but is pointing only the working example(so config.allow wide subnet and more specific on config.deny), maybe i am wrong and is not supposted to be used that way but for my use-case could be useful

### Additional Details & Logs

- Kong version (`# kong version 2.2.1`)
- This is a clean kong installation on Ubuntu 20.04.2 LTS (Focal Fossa)
- Kong error logs, i don't see any particular error on the kong logs
- If it helps to make it more clear i can provide you a test url to see and validate the ip-restrictions with https://check-host.net IPs

**Comment by bungle** [comment]
@fnzv,

Yes, currently the plugin does only check if it matches. And `deny` takes a priority in that case. It does not check whether the match is more specific or not. Though I think that would be a nice feature to add.
